File: python/src/bursar/billing/postgres/repositories/payment.py
from __future__ import annotations

from typing import Literal

from bursar.billing.postgres.repositories.schemas import BillingPaymentRow
from bursar.credits.postgres.repositories._types import DbQuery
from bursar.credits.postgres.repositories._utils import (
    optional_mapping_row,
    require_identifier_result,
    validate_non_empty,
    validate_row,
)


class BillingPaymentRepository:
    def __init__(self, execute: DbQuery) -> None:
        self._execute = execute

    def upsert(
        self,
        provider: str,
        provider_payment_id: str,
        provider_invoice_id: str | None,
        user_id: str,
        amount_minor: int,
        tax_minor: int,
        currency: str,
        purpose: Literal["subscription", "credit_topup"],
        metadata: str | None,
        status: Literal["pending", "succeeded", "failed", "canceled"],
        provider_updated_at: str,
    ) -> str:
        validate_non_empty(provider, "provider")
        validate_non_empty(provider_payment_id, "provider_payment_id")
        if not user_id:
            raise ValueError("billing payment requires a subject")
        if purpose not in {"subscription", "credit_topup"}:
            raise ValueError("billing payment requires a known purpose")
        rows = self._execute(
            "SELECT bursar.upsert_billing_payment("
            "%s::uuid,%s,%s,%s,%s,%s,%s,%s::bursar.billing_payment_status,%s,%s,%s::jsonb"
            ") AS id",
            [
                user_id,
                provider,
                provider_payment_id,
                amount_minor,
                tax_minor,
                currency,
                purpose,
                status,
                provider_updated_at,
                provider_invoice_id,
                metadata or "{}",
            ],
        )
        return require_identifier_result(rows, "id", "BillingPaymentRepository.upsert")

    def get_for_refund(self, provider: str, provider_payment_id: str) -> BillingPaymentRow | None:
        rows = self._execute(
            "SELECT * FROM bursar.get_billing_payment_by_provider(%s,%s)",
            [provider, provider_payment_id],
        )
        row = optional_mapping_row(rows, "BillingPaymentRepository.get_for_refund")
        if row is None:
            return None
        return validate_row(
            BillingPaymentRow,
            {
                "id": row.get("id"),
                "provider": row.get("provider"),
                "provider_payment_id": row.get("provider_payment_id"),
                "provider_invoice_id": row.get("provider_invoice_id"),
                "subject_id": row.get("subject_id"),
                "amount_minor": row.get("amount_minor"),
                "tax_minor": row.get("tax_minor"),
                "currency": row.get("currency"),
                "purpose": row.get("purpose"),
                "status": row.get("status"),
                "provider_updated_at": row.get("provider_updated_at"),
                "metadata": row.get("metadata"),
            },
            "BillingPaymentRepository.get_for_refund",
        )

    def get_direct(self, provider: str, provider_payment_id: str) -> BillingPaymentRow | None:
        return self.get_for_refund(provider, provider_payment_id)

File: javascript/src/billing/postgres/repositories/payment.ts
import { z } from "zod";
import type { QueryFn } from "../../../shared/postgres-types.js";
import type { PostgresRow } from "../../../shared/json.js";
import {
  optionalRecordRow,
  postgresUuid,
  requireResultField,
  safeParse,
} from "../../../shared/postgres-validation.js";

const PgSafeMinorUnitsSchema = z.union([
  z.string().regex(/^\d+$/),
  z.number().int().nonnegative().safe(),
]);

const BillingPaymentRowSchema = z
  .object({
    id: postgresUuid,
    provider: z.string().min(1),
    provider_payment_id: z.string().min(1),
    provider_invoice_id: z.string().nullable(),
    subject_id: postgresUuid,
    amount_minor: PgSafeMinorUnitsSchema,
    tax_minor: PgSafeMinorUnitsSchema,
    currency: z.string().regex(/^[A-Z]{3}$/),
    purpose: z.enum(["subscription", "credit_topup"]),
    status: z.enum(["pending", "succeeded", "failed", "canceled"]),
    provider_updated_at: z.union([z.string().datetime({ offset: true }), z.date()]),
    metadata: z.record(z.string(), z.json()),
  })
  .strict();
export type BillingPaymentRow = z.infer<typeof BillingPaymentRowSchema>;
export type ForRefundRow = BillingPaymentRow;

function parsePaymentRow(row: PostgresRow, context: string): BillingPaymentRow {
  return safeParse(
    BillingPaymentRowSchema,
    {
      id: row.id,
      provider: row.provider,
      provider_payment_id: row.provider_payment_id,
      provider_invoice_id: row.provider_invoice_id,
      subject_id: row.subject_id,
      amount_minor: row.amount_minor,
      tax_minor: row.tax_minor,
      currency: row.currency,
      purpose: row.purpose,
      status: row.status,
      provider_updated_at: row.provider_updated_at,
      metadata: row.metadata,
    },
    context,
  );
}

export class BillingPaymentRepository {
  constructor(private query: QueryFn) {}
  async upsert(
    provider: string,
    providerPaymentId: string,
    providerInvoiceId: string | null,
    userId: string,
    amountMinor: number,
    taxMinor: number,
    currency: string,
    purpose: "subscription" | "credit_topup",
    metadata: string | null,
    status: "pending" | "succeeded" | "failed" | "canceled",
    providerUpdatedAt: string,
  ): Promise<string> {
    if (!userId) throw new TypeError("billing payment requires a subject");
    if (purpose !== "subscription" && purpose !== "credit_topup") {
      throw new TypeError("billing payment requires a known purpose");
    }
    const rows = await this.query(
      `SELECT bursar.upsert_billing_payment(
         $1::uuid,$2,$3,$4,$5,$6,$7,$8::bursar.billing_payment_status,$9,$10,$11::jsonb
       ) AS id`,
      [
        userId,
        provider,
        providerPaymentId,
        amountMinor,
        taxMinor,
        currency,
        purpose,
        status,
        providerUpdatedAt,
        providerInvoiceId,
        metadata ?? "{}",
      ],
    );
    return requireResultField(rows, "id", postgresUuid, "BillingPaymentRepository.upsert");
  }
  async getForRefund(provider: string, providerPaymentId: string): Promise<ForRefundRow | null> {
    const rows = await this.query("SELECT * FROM bursar.get_billing_payment_by_provider($1,$2)", [
      provider,
      providerPaymentId,
    ]);
    const row = optionalRecordRow(rows, "BillingPaymentRepository.getForRefund");
    return row === null ? null : parsePaymentRow(row, "BillingPaymentRepository.getForRefund");
  }
  async getDirect(provider: string, providerPaymentId: string): Promise<BillingPaymentRow | null> {
    const rows = await this.query("SELECT * FROM bursar.get_billing_payment_by_provider($1,$2)", [
      provider,
      providerPaymentId,
    ]);
    const row = optionalRecordRow(rows, "BillingPaymentRepository.getDirect");
    return row === null ? null : parsePaymentRow(row, "BillingPaymentRepository.getDirect");
  }
}
